atomicsabi64/atomicsabi64.rst

```diff
@@ -0,0 +1,1252 @@
+
+
+We do not list other variants of ``fetch_<op>`` since their Mappings should be
+the same (modulo implementations of <op> that are not in scope of this
+document). Precisely, implementations that use loops should use the instructions
+that load or store from memory with the relevant memory order, and the
+appropriate <op> Assembly Sequence inside the loop. Exceptions, where Assembly 
+Sequences exist, are stated (for instance ``fetch_or`` can be implemented using
+``LDSETP`` when the LSE128 extension is enabled).
+
+Special Cases
+-------------
+
+There are special cases in the Mappings presented above, these must be handled
+in order to prevent unexpected outcomes of the compiled program. The special 
+cases are identified below.
+
+* Re-Ordering of Read-Modify-Write Effects and Acquire Fence
+* Const-Qualified 128-bit Atomic Loads
+
+Destination Register Should Not Be Zero Register for Read-Modify-Writes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A compiler is not permitted to rewrite the destination register to be the
+zero register for atomic operations that make use of ``SWP`` and ``LD<OP>``
+Assembly instructions. These include but are not limited to:
+
+.. table::
+
+  +-----------------------------------------+--------------------------------------+
+  | Atomic Operation                        | Assembly Sequence                    |
+  +=========================================+======================================+
+  | ``exchange(loc,val,sc)``                | ``MOV W4, #val;``                    |
+  |                                         | ``SWP W4, W10, [X1]``                |
+  +-----------------------------------------+--------------------------------------+
+  | ``fetch_add(loc,val,sc)``               | ``MOV W4, #val;``                    |
+  |                                         | ``LDADD W4, W10, [X1]``              |
+  +-----------------------------------------+--------------------------------------+
+
+Where ``X1`` contains the address of ``loc``.
+
+We annotate Mappings affected with ``*`` in section 4.2.
+
+Const-Qualified 128-bit Atomic Loads Should Be Marked Mutable
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Const-qualified data containing 128-bit atomic types should not be placed
+in read-only memory (such as the ``.rodata`` section).
+
+Before LSE2, the only way to implement a single-copy 128-bit atomic load
+is by using a Read-Modify-Write sequence. The write is not visible to
+software if the memory is writeable. Compilers and runtimes should use the
+LSE2/LRCPC3 sequence when available.
+
+
+Declarative statement of Mappings compatibility
+===============================================
+
+To ensure that the above Mappings are ABI-compatible we tested the compilation of
+Concurrent Programs, where each Atomic Operation is compiled to one of the
+aforementioned Mappings. We test if there is a compiled program that exhibits
```

Comment:
tested/test. We're switching tense. I understand that the sentences express different things, but it still reads as odd.